src/workerd/jsg/url.c++

```diff
@@ -394,4 +401,1853 @@ kj::Maybe<UrlSearchParams::EntryIterator::Entry> UrlSearchParams::EntryIterator:
   };
 }
 
+// ======================================================================================
+// UrlPattern
+
+namespace {
+
+constexpr auto MODIFIER_OPTIONAL = "?"_kjc;
+constexpr auto MODIFIER_ZERO_OR_MORE = "*"_kjc;
+constexpr auto MODIFIER_ONE_OR_MORE = "+"_kjc;
+
+inline bool isAsciiDigit(char c) { return c >= '0' && c <= '9'; };
+
+inline bool isHexDigit(char c) {
+  // Check if `c` is the ASCII code of a hexadecimal digit.
+  return isAsciiDigit(c) || ('a' <= c && c <= 'f') || ('A' <= c && c <= 'F');
+}
+
+inline bool isAscii(char codepoint) { return codepoint >= 0x00 && codepoint <= 0x7f; };
+
+inline bool isForbiddenHostCodepoint(char c) {
+  return c == 0x00 || c == 0x09 /* Tab */ || c == 0x0a /* LF */ || c == 0x0d /* CR */ ||
+         c == ' ' || c == '#' || c == '%' || c == '/' || c == ':' ||
```

Comment:
https://url.spec.whatwg.org/#forbidden-host-code-point doesn't have `%` in the forbidden host code point definition.

Reply:
This check is only used to work around an inconsistency/bug in the URLPattern spec in how it expects to handle hostnames. It is not used in the actual parsing of URLs. Look at the comment for `isValidHostnameInput` on line 438.